File: src/main/java/hzt/aoc/day24/Part2LobbyLayout.java
package hzt.aoc.day24;

import hzt.aoc.*;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

// Credits to Johan de Jong
public class Part2LobbyLayout extends Day24Challenge {

    public Part2LobbyLayout() {
        super("part 2",
                "How many tiles will be black after 100 days?");
    }

    private static final int DAYS_OF_EXHIBIT = 100;

    @Override
    protected long calculateResult(final List<List<String>> instructions) {
        final Map<Point2D, Tile> tileMap = buildFloorByInstructions(instructions);
        final Set<Tile> blackTiles = tileMap.values().stream().filter(Tile::isBlackUp).collect(Collectors.toSet());
        for (int day = 0; day < DAYS_OF_EXHIBIT; day++) {
            simulate(blackTiles);
        }
        return blackTiles.size();
    }

    private void simulate(final Set<Tile> blackTiles) {
        final Set<Tile> active = determineActiveSet(blackTiles);
        final Set<Tile> originalBlack = new HashSet<>(blackTiles);
        for (final Tile position : active) {
            final long blackNeighbours = countBlackNeighbours(originalBlack, position);
            if (originalBlack.contains(position) && (blackNeighbours == 0 || blackNeighbours > 2)) {
                blackTiles.remove(position);
            } else if (blackNeighbours == 2) {
                blackTiles.add(position);
            }
        }
    }

    private long countBlackNeighbours(final Set<Tile> originalBlack, final Tile startTile) {
        return startTile.neighbors().stream()
                .filter(originalBlack::contains)
                .count();
    }

    private Set<Tile> determineActiveSet(final Set<Tile> blackTiles) {
        final Set<Tile> active = new HashSet<>(blackTiles);
        for (final Tile position : blackTiles) {
            active.addAll(position.neighbors());
        }
        return active;
    }

    @Override
    String getMessage(final long global) {
        return String.format("%d", global);
    }
}

File: src/main/kotlin/hzt/aoc/day24/Part2LobbyLayout.kt
package hzt.aoc.day24

import java.util.stream.Collectors

// Credits to Johan de Jong
class Part2LobbyLayout : Day24Challenge(
    "part 2",
    "How many tiles will be black after 100 days"
) {
    override fun calculateResult(instructions: List<List<String>>): Long {
        val tileMap = buildFloorByInstructions(instructions)
        val blackTiles = tileMap.values.stream().filter { obj: Tile -> obj.isBlackUp }.collect(Collectors.toSet())
        for (day in 0 until DAYS_OF_EXHIBIT) {
            simulate(blackTiles)
        }
        return blackTiles.size.toLong()
    }

    private fun simulate(blackTiles: MutableSet<Tile>) {
        val active = determineActiveSet(blackTiles)
        val originalBlack: Set<Tile> = HashSet(blackTiles)
        for (position in active) {
            val blackNeighbours = countBlackNeighbours(originalBlack, position)
            if (originalBlack.contains(position) && (blackNeighbours == 0L || blackNeighbours > 2)) {
                blackTiles.remove(position)
            } else if (blackNeighbours == 2L) {
                blackTiles.add(position)
            }
        }
    }

    private fun countBlackNeighbours(originalBlack: Set<Tile>, startTile: Tile): Long {
        return startTile.neighbors().stream()
            .filter { originalBlack.contains(it) }
            .count()
    }

    private fun determineActiveSet(blackTiles: Set<Tile>): Set<Tile> {
        val active: MutableSet<Tile> = HashSet(blackTiles)
        for (position in blackTiles) {
            active.addAll(position.neighbors())
        }
        return active
    }

    override fun getMessage(value: Long): String {
        return String.format("%d", value)
    }

    companion object {
        private const val DAYS_OF_EXHIBIT = 100
    }
}
